### PR TITLE
Profile_Bears: Add Profiling Interface doc

### PR DESCRIPTION
**Pull request by RaiVaibhav**
Related to https://github.com/coala/coala/issues/565

### CODE REVIEW

File: docs/Developers/Profile_Bears.rst [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+This document provide the overview of coala's Profiling Interface. The
+Profiling Interface will provide an ability to Profile the Bear's code to
+optimize its performance.
+
+The profiler will start by profiling the run() method of bears because it
+consumes most of the bears time. So, this is part where bear writer will spend
+time, as rest of the part like loading the files, collecting the settings,
+etc. are done by coala itself.
+
+.. note::
```
Comment (Makman2) [inline docs/Developers/Profile_Bears.rst]: shouldn't be there another newline here? Not sure if RST requires that

File: docs/Developers/Profile_Bears.rst [hunk 2]
```diff
@@ -0,0 +1,74 @@
+Profile Bears
+=============
+
+This document provide the overview of coala's Profiling Interface. The
```
Comment (Makman2) [inline docs/Developers/Profile_Bears.rst]: provide**s** ;) --> https://www.gymglish.com/en/english-grammar/the-s-in-the-third-person-singular-form :D

Comment (Makman2) [inline docs/Developers/Profile_Bears.rst]: an overview

File: docs/Developers/Profile_Bears.rst [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+Profiling Interface will provide an ability to Profile the Bear's code to
```
Comment (Makman2) [inline docs/Developers/Profile_Bears.rst]: I think plural sounds better:

The Profiling Interfaces provides abilities to ~~P~~profile Bear~~s~~ code

File: docs/Developers/Profile_Bears.rst [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+optimize its performance.
+
+The profiler will start by profiling the run() method of bears because it
+consumes most of the bears time. So, this is part where bear writer will spend
```
Comment (Makman2) [inline docs/Developers/Profile_Bears.rst]: You should omit the part "consumes most the bears time", that's an assumption that is usually satisfied for larger setups.
Also you actually describe the real reason below :)
-->
The profiler will start by profiling the run() method of bears because ~~it
consumes most of the bears time. So,~~ this is *the* part where bear writer**s** will spend time on ...

File: docs/Developers/Profile_Bears.rst [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+time, as rest of the part like loading the files, collecting the settings,
+etc. are done by coala itself.
+
+.. note::
+    Enabliling both profiler and debugger on a bear at a same time will result
+    into error.
+
+Bear writer will have the ability to directly dump the raw profile output
```
Comment (Makman2) [inline docs/Developers/Profile_Bears.rst]: writer*s*

Comment (Makman2) [reply docs/Developers/Profile_Bears.rst]: Since this is an "unspecified quantity" of people you are referring two, you have to choices in fact:
1. A bear writer
2. Bear writers

File: docs/Developers/Profile_Bears.rst (hunk 1 above)
Comment (RaiVaibhav) [reply docs/Developers/Profile_Bears.rst]: Both will work fine IMO.

File: docs/Developers/Profile_Bears.rst [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+performance.
+
+The profiler will start by profiling the run() method of bears because this is
+the part where bear writers will spend time on, as rest of the part like loading
+the files, collecting the settings, etc. are done by coala itself.
+
+.. note::
+
+    Enabliling both profiler and debugger on a bear at a same time will result
+    into error.
```
Comment (Makman2) [inline docs/Developers/Profile_Bears.rst]: in~~to~~ *an* error.

File: docs/Developers/Profile_Bears.rst [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+Bear writers will have the ability to directly dump the raw profile output
+(binary output) either on current working directory or to a specified directory
```
Comment (Makman2) [inline docs/Developers/Profile_Bears.rst]: > (binary output)

Is that relevant?

File: docs/Developers/Profile_Bears.rst [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+name, which can be further used for examination of profiler stats with the help
+of different modules like ``pstats`` or ``snakeviz``.
+
+``--profile`` or ``profile`` (using coafile) is the main argument to enable
+profiling. It accepts an additional parameter directory name or directory path
+where profiled files will be dumped, if ``True`` is provided profiled files will
+be saved to current working directory. If specified directory does not exists it
+will create one and if path specified is already a file path then it will throw
+``FileExitsError``.
```
Comment (Makman2) [inline docs/Developers/Profile_Bears.rst]: Typo ;)

Comment (Makman2) [inline docs/Developers/Profile_Bears.rst]: I guess all this belongs into the `Usage` section. It should be maybe stripped as well since you describe already how to enable the profiler below. But the `FileExistsError` is important, that could go directly below the `Usage` paragraph, describing how paths are treated.

File: docs/Developers/Profile_Bears.rst [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+
+.. note::
+
+    All bears (even implicit dependency bears) in a section will be profiled.
+    Profiled files will have a name format {section.name}_{bear.name}.prof.
```
Comment (Makman2) [inline docs/Developers/Profile_Bears.rst]: should be enclosed in backticks ;)
```
``{section.name}...prof``
```

File: docs/Developers/Profile_Bears.rst [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+
+Usage
+-----
+
+Command Line Interface
+^^^^^^^^^^^^^^^^^^^^^^
+
+Bear writers can invoke the profiler with the ``--profile`` argument, i.e.
+
+To dump profiled files in current working directory.
```
Comment (Makman2) [inline docs/Developers/Profile_Bears.rst]: i.e. ?
Those sentences do not connect very well :)

File: docs/Developers/Profile_Bears.rst [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+
+.. code:: shell
+
+    $coala -b PEP8Bear,PyUnusedCodeBear -f <filename> --profile
+
+Or to dumped profiled files to a specified directory.
```
Comment (Makman2) [inline docs/Developers/Profile_Bears.rst]: to dump~~ed~~ profile~~d~~ files to a specified directory: (<-- note the colon)

File: docs/Developers/Profile_Bears.rst [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+
+.. code:: shell
+
+    $coala -b PEP8Bear,MypyBear -f <filename> --profile <Dirname/Dirpath>
```
Comment (Makman2) [inline docs/Developers/Profile_Bears.rst]: choose one, dirname or dirpath, as well as write it completely lower-case to match the same style as `<filename>`

File: docs/Developers/Profile_Bears.rst [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+
+coafile
+^^^^^^^
+
+Users can specify to profile bears using a ``.coafile``:
```
Comment (Makman2) [inline docs/Developers/Profile_Bears.rst]: I would add "as well":

> Users can specify to profile bears using a ``.coafile`` as well:

File: docs/Developers/Profile_Bears.rst [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+
+To dump profiled files in current working directory.
```
Comment (Makman2) [inline docs/Developers/Profile_Bears.rst]: Same here, sentences do not connect

File: docs/Developers/Profile_Bears.rst [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+
+::
+
+    [all]
+    bears = PEP8Bear,MypyBear
+    files = <filename>
+    profile = True
+
+Or to dumped profiled files to a specified directory.
```
Comment (Makman2) [inline docs/Developers/Profile_Bears.rst]: dump~~ed~~ profile~~d~~ ...: (<-- note the colon again)

File: docs/Developers/Profile_Bears.rst [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+be further used for examination of profiler stats with the help of different
+modules like ``pstats`` or ``snakeviz``.
+
+Usage
+-----
+
+coala's Profiler accepts an additional parameter, directory name or directory
+path where profiled files will be dumped, if ``True`` is provided profiled
+files will be saved to current working directory. If specified directory does
+not exists it will create one and if path specified is already a file path then
```
Comment (Makman2) [inline docs/Developers/Profile_Bears.rst]: If *the* specified directory does not exist~~s~~ it will be created ~~create one~~. If the specified path points to an already existing file a `FileExistsError` is thrown.

File: docs/Developers/Profile_Bears.rst [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+      profile = True
+
+- To dump profile files to a specified directory:
+
+  ::
+
+      [all]
+      bears = PEP8Bear,MypyBear
+      files = <filename>
+      profile = DirName/Dirpath
```
Comment (Makman2) [inline docs/Developers/Profile_Bears.rst]: dirname or dirpath, choose wisely :P

File: docs/Developers/Profile_Bears.rst [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+Bear writers will have the ability to directly dump the raw profile output
+either on current working directory or to a specified directory name, which can
+be further used for examination of profiler stats with the help of different
+modules like ``pstats`` or ``snakeviz``.
+
+Usage
+-----
+
+coala's Profiler accepts an additional parameter, directory name or directory
+path where profiled files will be dumped, if ``True`` is provided profiled
```
Comment (Makman2) [inline docs/Developers/Profile_Bears.rst]: When reading this sounds a bit confusing, it reads like it takes either "an additional parameter", or either a "directory name" or a "directory path". Would recommend following:
1. ... accepts an additional parameter: a directory path where profiled files will be ...

File: docs/Developers/Profile_Bears.rst [hunk 19]
```diff
@@ -0,0 +1,74 @@
...
+files will be saved to current working directory. If the specified directory
+does not exist it will be created. If the specified path points to an already
+existing file a ``FileExistsError`` is thrown.
```
Comment (Makman2) [inline docs/Developers/Profile_Bears.rst]: question is what that means to the regular user, but let's put that aside for now, I'm not sure how coala should handle such "user-exceptions" properly (which are intended to generate an error-log and exit).

**Comment by Makman2** [comment]
ack a5bf92418df9dff886fe91b3386e649d898f7895

**Comment by jayvdb** [comment]
As https://github.com/coala/coala/issues/565 is still open, I guess this could be `Closes ...` ?

File: docs/Developers/Profile_Bears.rst [hunk 20]
```diff
@@ -0,0 +1,76 @@
...
+Bear writers will have the ability to directly dump the raw profile output
+either on current working directory or to a specified directory name, which can
+be further used for examination of profiler stats with the help of different
+modules like ``pstats`` or ``snakeviz``.
+
+Usage
+-----
+
+coala's Profiler accepts an additional parameter: a directory path where
+profiled files will be dumped, if ``True`` is provided profiled files will be
```
Comment (jayvdb) [inline docs/Developers/Profile_Bears.rst]: `:` on line above is incompatible with the `,` here.

If you use a colon, you are allowed one phrase after that.  No more.

Then start a new sentence.

File: docs/Developers/Profile_Bears.rst [hunk 21]
```diff
@@ -0,0 +1,76 @@
...
+saved to current working directory. If the specified directory does not exist
+it will be created. If the specified path points to an already existing file
+a error is raised. Profiled files will get overwritten, if specified directory
```
Comment (jayvdb) [inline docs/Developers/Profile_Bears.rst]: > Profiled files will get overwritten, if specified directory

I thought this was will cause an error?

If I am correct, you obviously have not re-checked this document and it shouldnt be up for review.

Comment (RaiVaibhav) [reply docs/Developers/Profile_Bears.rst]: `FileExistsError` will be raise when user provide a `filepath` in `--profile` but if user provide a `directory path` and that directory already have `profiled` files then in the next run of coala file with the same name gets overwritten

File: docs/Developers/Profile_Bears.rst [hunk 22]
```diff
@@ -0,0 +1,76 @@
...
+Bear writers will have the ability to directly dump the raw profile output
+either on current working directory or to a specified directory name, which can
+be further used for examination of profiler stats with the help of different
+modules like ``pstats`` or ``snakeviz``.
+
+Usage
+-----
+
+coala's Profiler accepts an additional parameter, a directory path where
+profiled files will be dumped, if ``True`` is provided profiled files will be
```
Comment (jayvdb) [inline docs/Developers/Profile_Bears.rst]: ill avoid the next grammar lesson.

After `dumped` add a full stop.  Add two carraige returns.

Short simple intros are what people will read.  If the short intro is stuffed into a large paragraph, they will try to avoid it.

Then explain the `profile` arg options in a new paragraph.  And this usage section should not mention "True" , the title case Python constant, as there are multiple ways to enable profiling, and many do not require the Python keyword `True`.  Dumb it down.  Use prose.

**Comment by jayvdb** [comment]
ack https://github.com/coala/coala/pull/5666/commits/0e075217ea1bc5732d9deac57355431db00ef4ee

**Comment by jayvdb** [comment]
@gitmate-bot ff
